Automated Smoketest: :x:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 75ed72fc83efd3b54f691e014b25a77a0aa6b3fa
Build: success
ECL Watch:  Rebuilding Site
src/GraphStore.ts(6,1): error TS1185: Merge conflict marker encountered.
src/GraphStore.ts(23,1): error TS1185: Merge conflict marker encountered.
src/GraphStore.ts(25,1): error TS1185: Merge conflict marker encountered.
CMake Warning at /mnt/disk1/home/vamosax/smoketest/PR-11631/build/esp/src/cmake_install.cmake:65 (message):
  npm run compile failed see above for details.
Call Stack (most recent call first):
  /mnt/disk1/home/vamosax/smoketest/PR-11631/build/esp/cmake_install.cmake:41 (include)
  /mnt/disk1/home/vamosax/smoketest/PR-11631/build/cmake_install.cmake:50 (include)
CPack: Create package

Install warning(s): 
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: phantomjs-prebuilt@2.1.16 (node_modules/jshint/node_modules/phantomjs-prebuilt):
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: phantomjs-prebuilt@2.1.16 install: `node install.js`
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: Exit status 1

Install hpccsystems-platform-community_7.0.0-rc1.el7.x86_64.rpm
Creating group hpcc ...\x1b[32m [   OK    ]  \x1b[0m 
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |94 |94 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |805 |805 |0 |
|test (thor) |729 |729 |0 |
|test (roxie) |878 |878 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

[Build error:](https://gist.github.com/9167eb4f008a1093b37ecac5623f94e9)
Build log [Build log](https://gist.github.com/HPCCSmoketest/9167eb4f008a1093b37ecac5623f94e9#file-result-18-09-06-01-02-13-log)

Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   34 sec (00:00:34) |  175 sec (00:02:55) |  180 sec (00:03:00) |   20 sec (00:00:20) |   42 sec (00:00:42) | 1355 sec (00:22:35) |   19 sec (00:00:19) | 1825 sec (00:30:25) |

--- a/esp/src/eclwatch/GraphsWidget.js
+++ b/esp/src/eclwatch/GraphsWidget.js
@@ -252,7 +252,7 @@ define([
                 this._onRowDblClick({ Name: graphName }, { SubGraphId: subgraphID });
             },
 
-            createDetail: function (id, row, params) {
+            createDetail: function (_id, row, params) {
                 params = params || {};
                 if (this.mode === "Query") {
                     params.legacyMode = true;
@@ -284,6 +284,7 @@ define([
                 var title = row.Name;
                 var delayWidget = "GraphTree7Widget";
                 var delayProps = {
+                    _owner: this,
                     forceJS: true
                 };
                 if (params && params.SubGraphId) {
@@ -295,7 +296,7 @@ define([
                     delayProps = {};
                 }
                 return new DelayLoadWidget({
-                    id: id,
+                    id: _id,
                     title: title,
                     closable: true,
                     delayWidget: delayWidget,

--- a/esp/src/eclwatch/MembersWidget.js
+++ b/esp/src/eclwatch/MembersWidget.js
@@ -188,7 +188,7 @@ define([
                 }).placeAt(this.dialog.domNode, "last");
 
                 on(this.dialogValidationTextBox, "keyup", function(event){
-                    if (context.dialogValidationTextBox.get("value") != "") {
+                    if (context.dialogValidationTextBox.get("value") !== "") {
                         context.dialogButton.set("disabled", false);
                     } else {
                         context.dialogButton.set("disabled", true);

--- a/esp/src/eclwatch/TimingPageWidget.js
+++ b/esp/src/eclwatch/TimingPageWidget.js
@@ -5,7 +5,6 @@ define([
     "dojo/i18n!./nls/hpcc",
     "dojo/_base/array",
     "dojo/on",
-    "dojo/store/Memory",
     "dojo/store/Observable",
 
     "dijit/registry",
@@ -33,9 +32,13 @@ define([
     "dijit/Toolbar",
     "dijit/ToolbarSeparator",
     "dijit/form/Button",
-    "dijit/form/Select"
+    "dijit/form/Select",
+    "dijit/form/MultiSelect",
+    "dijit/form/DropDownButton",
+    "dijit/TooltipDialog"
 
-], function (declare, lang, i18n, nlsHPCC, arrayUtil, on, Memory, Observable,
+
+], function (declare, lang, i18n, nlsHPCC, arrayUtil, on, Observable,
     registry, BorderContainer, TabContainer, ContentPane,
     selector,
     _TabContainerWidget, ESPWorkunit, DelayLoadWidget, ESPUtil, srcTimings,
@@ -103,8 +106,11 @@ define([
             },
 
             _onMetricsType: function (evt) {
-                this._metricFilter = evt.target.value;
+                this._metricFilter = this._timings.selectedMetrics();
                 this.refreshGrid();
+            },
+
+            _onMetricsClose: function (evt) {
             },
 
             _onReset: function () {
@@ -121,7 +127,7 @@ define([
                     context.refreshGrid();
                 }
 
-                var store = new Memory({
+                var store = new ESPUtil.UndefinedMemory({
                     idProperty: "__hpcc_id",
                     data: []
                 });
@@ -152,32 +158,44 @@ define([
 
             refreshGrid: function (forceRefresh) {
                 var context = this;
-                this._timings.refresh(forceRefresh).then(function (data) {
-                    context.grid.set("columns", {
+                this._timings.refresh(forceRefresh).then(function (args) {
+                    var cols = args[0];
+                    var data = args[1];
+                    var columns = {
                         __hpcc_id: { label: "##", width: 45 },
                         Name: {
                             label: context.i18n.Name,
                             sortable: true,
-                            formatter: function (Name, row) {
+                            width: 120,
+                            formatter: function (cell, row) {
                                 switch (row.Type) {
                                     case "graph":
                                     case "subgraph":
                                     case "activity":
                                     case "edge":
-                                        return "<a href='#" + Name + "' class='dgrid-row-url'>" + Name + "</a>";
+                                        return "<a href='#" + cell + "' class='dgrid-row-url'>" + cell + "</a>";
                                 }
-                                return Name;
+                                return cell;
                             }
-                        },
-                        Seconds: {
-                            label: context._timings._metricSelectLabel,
-                            width: 240
+                        }
+                    };
+                    cols.forEach(function (col) {
+                        var formattedID = "__" + col;
+                        columns[col] = {
+                            label: col,
+                            width: 120,
+                            formatter: function (cell, row) {
+                                var retVal = row[formattedID] || cell;
+                                return retVal !== undefined ? retVal : "";
+                            }
                         }
                     });
+                    context.grid.set("columns", columns);
                     context.store.setData(data.map(function (row, i) {
                         var GraphName;
                         var SubGraphId;
                         var ActivityId;
+                        var EdgeId;
                         switch (row.type) {
                             case "graph":
                                 GraphName = row.id;
@@ -187,22 +205,33 @@ define([
                                 SubGraphId = row.id;
                                 break;
                             case "activity":
-                            case "edge":
                                 GraphName = context._timings.graphID(row.name);
                                 SubGraphId = context._timings.subgraphID(row.name);
                                 ActivityId = row.id;
                                 break;
+                            case "edge":
+                                GraphName = context._timings.graphID(row.name);
+                                SubGraphId = context._timings.subgraphID(row.name);
+                                EdgeId = row.id;
+                                break;
                         }
-                        return {
+                        var dataRow = {
                             __hpcc_id: i + 1,
                             Name: row.id,
-                            Seconds: row[context._timings._metricSelectValue],
                             Type: row.type,
                             Scope: row.name,
                             GraphName: GraphName,
                             SubGraphId: SubGraphId,
-                            ActivityId: ActivityId
+                            ActivityId: ActivityId,
+                            EdgeId: EdgeId
+                        };
+                        for (var key in row){
+                            dataRow[key] = row[key];
                         }
+                        cols.forEach(function(col) {
+                            dataRow[col] = row[col];
+                        });
+                        return dataRow;
                     }));
                     context.grid.refresh();
                 });
@@ -220,20 +249,34 @@ define([
                 }
             },
 
-            ensurePane: function (id, params) {
+            openGraph: function(graphName, subgraphID) {
+                var tab = this.ensurePane(subgraphID + " - " + graphName, {
+                    GraphName: graphName,
+                    SubGraphId: subgraphID
+                });
+                if (tab) {
+                    this.selectChild(tab);
+                }
+            },
+
+            ensurePane: function (_id, params) {
+                var id = this.createChildTabID(_id);
                 var retVal = registry.byId(id);
                 if (!retVal) {
-                    var context = this;
                     retVal = new DelayLoadWidget({
-                        id: this.createChildTabID(id),
-                        title: id,
+                        id: id,
+                        title: _id,
                         closable: true,
                         delayWidget: "GraphTree7Widget",
+                        delayProps: {
+                            _owner: this
+                        },
                         params: {
                             Wuid: this.wu.Wuid,
                             GraphName: params.GraphName,
                             SubGraphId: params.SubGraphId,
-                            ActivityId: params.ActivityId
+                            ActivityId: params.ActivityId,
+                            EdgeId: params.EdgeId
                         }
                     });
                     if (retVal) {

--- a/esp/src/eclwatch/dojoConfig.js
+++ b/esp/src/eclwatch/dojoConfig.js
@@ -24,6 +24,7 @@ function getConfig(env) {
             "templates": baseUrl + "/eclwatch/templates",
             "ecl": baseUrl + "/eclwatch/ecl",
             "css": baseUrl + "/loader/css",
+            "d3-array": baseUrl + "/node_modules/d3-array/build/d3-array",
             "d3-selection": baseUrl + "/node_modules/d3-selection/build/d3-selection",
             "@hpcc-js/api": baseUrl + "/node_modules/@hpcc-js/api/dist/index" + hpccMin,
             "@hpcc-js/chart": baseUrl + "/node_modules/@hpcc-js/chart/dist/index" + hpccMin,

--- a/esp/src/eclwatch/nls/hpcc.js
+++ b/esp/src/eclwatch/nls/hpcc.js
@@ -859,6 +859,8 @@ define({root:
     Zip: "Zip",
     ZippedAnalysisPackage: "Zipped Analysis Package",
     Zoom: "Zoom",
+    ZoomPlus: "Zoom +",
+    ZoomMinus: "Zoom -",
     Zoom100Pct: "Zoom 100%",
     ZoomAll: "Zoom All",
     ZoomWidth: "Zoom Width"

--- a/esp/src/eclwatch/stub.js
+++ b/esp/src/eclwatch/stub.js
@@ -37,6 +37,16 @@ define([
         var _prevReset = Date.now();
         var sessionIsActive = cookie("ESPSessionTimeoutSeconds");
 
+        function _resetESPTime(evt) {
+            if (Date.now() - _prevReset > SESSION_RESET_FREQ) {
+                _prevReset = Date.now();
+                xhr("esp/reset_session_timeout", {
+                    method: "post"
+                }).then(function (data) {
+                });
+            }
+        }
+
         if (sessionIsActive > -1) {
             cookie("Status", "Unlocked");
             cookie("ECLWatchUser", "true");
@@ -62,16 +72,6 @@ define([
             monitorLockClick.unlocked();
         } else if (cookie("ECLWatchUser")) {
             window.location.replace(dojoConfig.urlInfo.basePath + "/Login.html");
-        }
-
-        function _resetESPTime(evt) {
-            if (Date.now() - _prevReset > SESSION_RESET_FREQ) {
-                _prevReset = Date.now();
-                xhr("esp/reset_session_timeout", {
-                    method: "post"
-                }).then(function (data) {
-                });
-            }
         }
 
         function startLoading(targetNode) {

--- a/esp/src/eclwatch/templates/GraphTree7Widget.html
+++ b/esp/src/eclwatch/templates/GraphTree7Widget.html
@@ -17,13 +17,13 @@
                 <div id="${id}GraphToolbar" class="topPanel dijit dijitToolbar" role="toolbar">
                     <div data-dojo-attach-event="onClick:_onGraphRefresh" data-dojo-props="iconClass:'fa fa-refresh', showLabel:false" data-dojo-type="dijit.form.Button">${i18n.Refresh}</div>
                     <span data-dojo-type="dijit.ToolbarSeparator"></span>
-                    <div data-dojo-attach-event="onClick:_onPartial" data-dojo-props="iconClass:'fa fa-window-restore', showLabel:false" data-dojo-type="dijit.form.Button">${i18n.Refresh}</div>
-                    <div data-dojo-attach-event="onClick:_onMax" data-dojo-props="iconClass:'fa fa-window-maximize', showLabel:false" data-dojo-type="dijit.form.Button">${i18n.Refresh}</div>
+                    <div data-dojo-attach-event="onClick:_onPartial" data-dojo-props="iconClass:'fa fa-window-restore', showLabel:false" data-dojo-type="dijit.form.Button">${i18n.CollapseAll}</div>
+                    <div data-dojo-attach-event="onClick:_onMax" data-dojo-props="iconClass:'fa fa-window-maximize', showLabel:false" data-dojo-type="dijit.form.Button">${i18n.Restore}</div>
                     <span data-dojo-type="dijit.ToolbarSeparator"></span>
-                    <div data-dojo-attach-event="onClick:_onZoomToFit" data-dojo-props="iconClass:'fa fa-arrows-alt', showLabel:false" data-dojo-type="dijit.form.Button">${i18n.Refresh}</div>
-                    <div data-dojo-attach-event="onClick:_onZoomToWidth" data-dojo-props="iconClass:'fa fa-arrows-h', showLabel:false" data-dojo-type="dijit.form.Button">${i18n.Refresh}</div>
-                    <div data-dojo-attach-event="onClick:_onZoomToPlus" data-dojo-props="iconClass:'fa fa-plus', showLabel:false" data-dojo-type="dijit.form.Button">${i18n.Refresh}</div>
-                    <div data-dojo-attach-event="onClick:_onZoomToMinus" data-dojo-props="iconClass:'fa fa-minus', showLabel:false" data-dojo-type="dijit.form.Button">${i18n.Refresh}</div>
+                    <div data-dojo-attach-event="onClick:_onZoomToFit" data-dojo-props="iconClass:'fa fa-arrows-alt', showLabel:false" data-dojo-type="dijit.form.Button">${i18n.ZoomAll}</div>
+                    <div data-dojo-attach-event="onClick:_onZoomToWidth" data-dojo-props="iconClass:'fa fa-arrows-h', showLabel:false" data-dojo-type="dijit.form.Button">${i18n.ZoomWidth}</div>
+                    <div data-dojo-attach-event="onClick:_onZoomToPlus" data-dojo-props="iconClass:'fa fa-plus', showLabel:false" data-dojo-type="dijit.form.Button">${i18n.ZoomPlus}</div>
+                    <div data-dojo-attach-event="onClick:_onZoomToMinus" data-dojo-props="iconClass:'fa fa-minus', showLabel:false" data-dojo-type="dijit.form.Button">${i18n.ZoomMinus}</div>
                     <span data-dojo-type="dijit.ToolbarSeparator"></span>
                     <span id="${id}GraphStatus" ></span>
                 </div>
@@ -56,13 +56,6 @@
                 <div id="${id}Properties" title="${i18n.Properties}" data-dojo-type="dijit.layout.ContentPane">
                 </div>
             </div>
-        </div>
-    </div>
-    <div id="${id}InfoDialog" title="${i18n.InfoDialog}" data-dojo-type="dijit.Dialog">
-        <div id="${id}InfoDialogContent" class="dijitDialogPaneContentArea">
-        </div>
-        <div class="dijitDialogPaneActionBar">
-            <button id="${id}InfoDialogCancel" type="button" data-dojo-type="dijit.form.Button">${i18n.Cancel}</button>
         </div>
     </div>
 </div>

--- a/esp/src/eclwatch/templates/TimingPageWidget.html
+++ b/esp/src/eclwatch/templates/TimingPageWidget.html
@@ -5,9 +5,19 @@
                 <div id="${id}Toolbar" class="topPanel" data-dojo-props="region: 'top'" data-dojo-type="dijit.Toolbar">
                     <div id="${id}Refresh" data-dojo-attach-event="onClick:_onRefresh" data-dojo-props="iconClass:'iconRefresh'" data-dojo-type="dijit.form.Button">${i18n.Refresh}</div>
                     <span data-dojo-type="dijit.ToolbarSeparator"></span>
-                    <select id="${id}MetricsType" style="box-shadow: none" data-dojo-attach-event="onChange:_onMetricsType">
-                        <option value="TimeElapsed">TimeElapsed</option>
-                    </select>
+                    <div id="${id}MetricsDropDown" data-dojo-type="dijit.form.DropDownButton">
+                        <span>Metrics</span>
+                        <div data-dojo-type="dijit.TooltipDialog">
+                            <div id="${id}MetricsForm" style="width: 240px;" onsubmit="return false;" data-dojo-type="dijit.form.Form">
+                                <select id="${id}MetricsType" multiple style="box-shadow: none;width:100%;height:320px" data-dojo-attach-event="onChange:_onMetricsType">
+                                    <option value="TimeElapsed">TimeElapsed</option>
+                                </select>
+                            </div>
+                            <div class="dijitDialogPaneActionBar">
+                                <button type="submit" data-dojo-attach-event="onClick:_onMetricsClose" data-dojo-type="dijit.form.Button">${i18n.Close}</button>
+                            </div>
+                        </div>
+                    </div>
                     <span data-dojo-type="dijit.ToolbarSeparator"></span>
                     <div id="${id}StackController" style="width: 100%" data-dojo-props="containerId:'${id}OverviewTabContainer'" data-dojo-type="dijit.layout.StackController"></div>
                     <span data-dojo-type="dijit.ToolbarSeparator"></span>

--- a/esp/src/src/ESPUtil.ts
+++ b/esp/src/src/ESPUtil.ts
@@ -11,19 +11,30 @@ import * as json from "dojo/json";
 import * as aspect from "dojo/aspect";
 import * as Evented from "dojo/Evented";
 import * as on from "dojo/on";
+import * as Memory from "dojo/store/Memory";
 
 import * as registry from "dijit/registry";
 import * as Tooltip from "dijit/Tooltip";
 
+// @ts-ignore
 import * as DGrid from "dgrid/Grid";
+// @ts-ignore
 import * as OnDemandGrid from "dgrid/OnDemandGrid";
+// @ts-ignore
 import * as Keyboard from "dgrid/Keyboard";
+// @ts-ignore
 import * as Selection from "dgrid/Selection";
+// @ts-ignore
 import * as ColumnResizer from "dgrid/extensions/ColumnResizer";
+// @ts-ignore
+import * as ColumnHider from "dgrid/extensions/ColumnHider";
+// @ts-ignore
 import * as DijitRegistry from "dgrid/extensions/DijitRegistry";
 
 import { select as d3Select } from "d3-selection";
 import { Pagination } from "./Pagination";
+
+import declareDecorator from './DeclareDecorator';
 
 declare const dojo;
 
@@ -309,6 +320,50 @@ export const FormHelper = declare(null, {
         return "";
     }
 });
+
+type Memory = {
+    data: any[];
+    queryEngine: any;
+    setData(data: any);
+    query(query: any, options: any): any;
+};
+
+export interface UndefinedMemoryBase extends Memory {
+}
+
+@declareDecorator("UndefinedMemoryBase", Memory)
+export class UndefinedMemoryBase { }
+
+export class UndefinedMemory extends UndefinedMemoryBase {
+
+    constructor() {
+        super();
+    }
+
+    query(query: any, options: any) {
+        if (options.sort !== undefined) {
+            if (options.sort.length > 0) {
+                var sortSet = options.sort;
+                options.sort = function (a, b) {
+                    for (var sort, i = 0; sort = sortSet[i]; i++) {
+                        var aValue = a[sort.attribute];
+                        var bValue = b[sort.attribute];
+                        // valueOf enables proper comparison of dates
+                        aValue = aValue != null ? aValue.valueOf() : aValue;
+                        bValue = bValue != null ? bValue.valueOf() : bValue;
+                        if (aValue != bValue) {
+                            if (aValue === undefined) return 1;     // Force "undefined" to bottom always
+                            if (bValue === undefined) return -1;    // Force "undefined" to bottom always
+                            return !!sort.descending == (aValue == null || aValue > bValue) ? -1 : 1;
+                        }
+                    }
+                    return 0;
+                }
+            }
+        }
+        return super.query(query, options);
+    }
+}
 
 export function Grid(pagination?, selection?, overrides?) {
     var baseClass = [];

--- a/esp/src/src/GraphStore.ts
+++ b/esp/src/src/GraphStore.ts
@@ -1,25 +1,9 @@
 import * as lang from "dojo/_base/lang";
 import * as arrayUtil from "dojo/_base/array";
-import * as Memory from "dojo/store/Memory";
 import * as QueryResults from "dojo/store/util/QueryResults";
+import {UndefinedMemory} from "./ESPUtil";
 
-import declareDecorator from './DeclareDecorator';
-
-type Memory = {
-    data: any[];
-    queryEngine: any;
-    setData(data: any);
-    query(query: any, options: any): any;
-};
-
-export interface GraphStoreBase extends Memory {
-}
-
-@declareDecorator("GraphStoreBase", Memory)
-export class GraphStoreBase { }
-
-export class GraphStore extends GraphStoreBase {
-
+export class GraphStore extends UndefinedMemory {
     idProperty: string;
     cacheColumns = {};
 

--- a/esp/src/src/GraphTree7Widget.ts
+++ b/esp/src/src/GraphTree7Widget.ts
@@ -63,10 +63,13 @@ export class GraphTree7Widget {
     GraphName: any;
 
     subGraphId: string;
+    activityId: string;
+    edgeId: string;
     graphTimers: any;
     targetQuery: any;
     queryId: any;
 
+    _owner;
     wuid = "";
     graphName = "";
     subgraphsGrid = null;
@@ -184,32 +187,7 @@ export class GraphTree7Widget {
         });
         grid.on(".dgrid-row:dblclick", function (evt) {
             var item = grid.row(evt).data;
-            if (item.Id) {
-                let refresh = false;
-                let scopeItem = context._gc.scopeItem(item.Id);
-                while (scopeItem) {
-                    const w = context._gc.item(scopeItem._.Id);
-                    if (w && w instanceof Subgraph && w.minState() !== "normal") {
-                        w.minState("normal");
-                        refresh = true;
-                    }
-                    scopeItem = scopeItem.parent;
-                }
-                const w = context._gc.item(item.Id);
-                if (refresh) {
-                    context._graph
-                        .data(context._gc.graphData(), true)   //  Force re-render 
-                        .render(w => {
-                            setTimeout(() => {
-                                context._graph
-                                    .centerOnItem(w)
-                                    ;
-                            }, 1000);
-                        });
-                } else {
-                    context._graph.centerOnItem(w);
-                }
-            }
+            context.centerOn(item.Id)
         });
     }
 
@@ -317,7 +295,7 @@ export class GraphTree7Widget {
 
     refresh(params) {
         if (params.SubGraphId) {
-            this.syncSelectionFrom([params.SubGraphId]);
+            this.syncSelectionFrom(this);
         }
     }
 
@@ -325,21 +303,33 @@ export class GraphTree7Widget {
         this.wuid = params.Wuid;
         this.graphName = params.GraphName;
         this.subGraphId = params.SubGraphId;
+        this.activityId = params.ActivityId;
+        this.edgeId = params.EdgeId;
         this.targetQuery = params.Target;
         this.queryId = params.QueryId;
 
-        this.refreshData();
+        this.refreshData().then(() => {
+            this.syncSelectionFrom(this);
+            if (this.edgeId) {
+                this.centerOn(this.edgeId);
+            } else if (this.activityId) {
+                this.centerOn(this.activityId);
+            } else if (this.subGraphId) {
+                this.centerOn(this.subGraphId);
+            }
+        });
     }
 
     refreshData() {
         if (this.isWorkunit()) {
-            this.loadGraphFromWu(this.wuid, this.graphName, this.subGraphId, true);
+            return this.loadGraphFromWu(this.wuid, this.graphName, this.subGraphId, true);
         } else if (this.isQuery()) {
         }
+        return Promise.resolve();
     }
 
     loadGraphFromWu(wuid, graphName, subGraphId, refresh: boolean = false) {
-        this.fetchScopeGraph(wuid, graphName, subGraphId, refresh).then(() => {
+        return this.fetchScopeGraph(wuid, graphName, subGraphId, refresh).then(() => {
             this.loadGraph();
             this.loadSubgraphs();
             this.loadVertices();
@@ -411,6 +401,11 @@ export class GraphTree7Widget {
         this.subgraphsGrid = new declare([Grid(true, true)])({
             store: this.subgraphsStore
         }, this.id + "SubgraphsGrid");
+        const context = this;
+        this.subgraphsGrid.on(".dgrid-row-url:click", function (evt) {
+            var row = context.subgraphsGrid.row(evt).data;
+            context._owner.openGraph(context.graphName, row.Id);
+        });
 
         this._initItemGrid(this.subgraphsGrid);
     }
@@ -418,12 +413,13 @@ export class GraphTree7Widget {
     loadSubgraphs() {
         var subgraphs = this._gc.subgraphStoreData();
         this.subgraphsStore.setData(subgraphs);
+        const context = this;
+        const img = Utility.getImageURL("folder.png");
         var columns = [
             {
                 label: this.i18n.ID, field: "Id", width: 54,
                 formatter: function (_id, row) {
-                    var img = Utility.getImageURL("folder.png");
-                    return "<img src='" + img + "'/>&nbsp;" + _id;
+                    return "<img src='" + img + "'/>&nbsp;" + (context._owner ? "<a href='#" + _id + "' class='dgrid-row-url'>" + _id + "</a>" : _id);
                 }
             }
         ];
@@ -480,6 +476,35 @@ export class GraphTree7Widget {
         this.edgesGrid.refresh();
     }
 
+    centerOn(itemID?: string) {
+        if (itemID) {
+            let refresh = false;
+            let scopeItem = this._gc.scopeItem(itemID);
+            while (scopeItem) {
+                const w = this._gc.item(scopeItem._.Id);
+                if (w && w instanceof Subgraph && w.minState() !== "normal") {
+                    w.minState("normal");
+                    refresh = true;
+                }
+                scopeItem = scopeItem.parent;
+            }
+            const w = this._gc.item(itemID);
+            if (refresh) {
+                this._graph
+                    .data(this._gc.graphData(), true)   //  Force re-render 
+                    .render(w => {
+                        setTimeout(() => {
+                            this._graph
+                                .centerOnItem(w)
+                                ;
+                        }, 1000);
+                    });
+            } else {
+                this._graph.centerOnItem(w);
+            }
+        }
+    }
+
     inSyncSelectionFrom = false;
     syncSelectionFrom(sourceControl) {
         if (!this.inSyncSelectionFrom) {
@@ -506,7 +531,7 @@ export class GraphTree7Widget {
         var tab = this.widget.OverviewTabContainer.get("selectedChildWidget");
         this.setDisabled(this.id + "FindPrevious", this.foundIndex <= 0, "iconLeft", "iconLeftDisabled");
         this.setDisabled(this.id + "FindNext", this.foundIndex >= this.found.length - 1, "iconRight", "iconRightDisabled");
-        this.setDisabled(this.id + "ActivityMetric", tab.id !== this.id + "ActivitiesTreeMap");
+        this.setDisabled(this.id + "ActivityMetric", tab && tab.id !== this.id + "ActivitiesTreeMap");
     }
 }
 
@@ -516,7 +541,15 @@ GraphTree7Widget.prototype._syncSelectionFrom = Utility.debounce(function (this:
     var selectedGlobalIDs = sourceControlOrGlobalIDs instanceof Array ? sourceControlOrGlobalIDs : [];
     if (sourceControl) {
         //  Get Selected Items  ---
-        if (sourceControl === this._graph) {
+        if (sourceControl === this) {
+            if (this.edgeId) {
+                selectedGlobalIDs = [this.edgeId];
+            } else if (this.activityId) {
+                selectedGlobalIDs = [this.activityId];
+            } else if (this.subGraphId) {
+                selectedGlobalIDs = [this.subGraphId];
+            }
+        } else if (sourceControl === this._graph) {
             selectedGlobalIDs = this._graph.selection()
                 .map((w: any) => this._gc.rItem(w))
                 .filter(item => !!item)

--- a/esp/src/src/Pagination.ts
+++ b/esp/src/src/Pagination.ts
@@ -3,7 +3,9 @@ import * as on from "dojo/on";
 
 // @ts-ignore
 import * as DGrid from "dgrid/Grid";
+// @ts-ignore
 import * as _StoreMixin from "dgrid/_StoreMixin";
+// @ts-ignore
 import * as DGridPagination from "dgrid/extensions/Pagination";
 
 export const Pagination = declare([DGridPagination], {

--- a/esp/src/src/Timings.ts
+++ b/esp/src/src/Timings.ts
@@ -2,6 +2,7 @@ import { Workunit } from "@hpcc-js/comms";
 import * as hpccCommon from "@hpcc-js/common";
 import { WUTimeline } from "@hpcc-js/eclwatch";
 import { Column } from "@hpcc-js/chart";
+import { ascending as d3Ascending } from "d3-array";
 
 const d3Select = (hpccCommon as any).select;
 
@@ -63,8 +64,15 @@ export class Timings {
         this.metricsSelect = d3Select(`#${metricsSelectTarget}`);
     }
 
-    init(wuid: string) {
+    selectedMetrics() {
+        this._metricSelectValue = this.metricsSelect.selectAll("option").nodes()
+            .filter(n => n.selected === true)
+            .map(n => n.value)
+            ;
+        return this._metricSelectValue;
+    }
 
+    init(wuid: string) {
     }
 
     protected walkScopeName(id: string, visitor: (name: string) => boolean) {
@@ -97,7 +105,7 @@ export class Timings {
 
     _scopeFilter: string = "";
     _metricSelectLabel: string = "";
-    _metricSelectValue: string = "";
+    _metricSelectValue: string[] = ["TimeElapsed"];
     _graphLookup: { [id: string]: string } = {};
     _subgraphLookup: { [id: string]: string } = {};
     fetchDetailsNormalizedPromise;
@@ -113,7 +121,7 @@ export class Timings {
                 ;
         }
         if (force || !this.fetchDetailsNormalizedPromise) {
-            this.fetchDetailsNormalizedPromise = this.wu.fetchDetailsNormalized({
+            this.fetchDetailsNormalizedPromise = Promise.all([this.wu.fetchDetailsMeta(), this.wu.fetchDetailsRaw({
                 ScopeFilter: {
                     MaxDepth: 999999,
                     ScopeTypes: []
@@ -141,20 +149,91 @@ export class Timings {
                     IncludeCreator: false,
                     IncludeCreatorType: false
                 }
+            })]).then(promises => {
+                const meta = promises[0];
+                const scopes = promises[1];
+                const columns: { [id: string]: any } = {
+                    id: {
+                        Measure: "label"
+                    },
+                    name: {
+                        Measure: "label"
+                    },
+                    type: {
+                        Measure: "label"
+                    }
+                };
+                const data: object[] = [];
+                for (const scope of scopes) {
+                    const props = {};
+                    if (scope && scope.Id && scope.Properties && scope.Properties.Property) {
+                        for (const key in scope.Properties.Property) {
+                            const scopeProperty = scope.Properties.Property[key];
+                            if (scopeProperty.Measure === "ns") {
+                                scopeProperty.Measure = "s";
+                            }
+                            columns[scopeProperty.Name] = { ...scopeProperty };
+                            delete columns[scopeProperty.Name].RawValue;
+                            delete columns[scopeProperty.Name].Formatted;
+                            switch (scopeProperty.Measure) {
+                                case "bool":
+                                    props[scopeProperty.Name] = !!+scopeProperty.RawValue;
+                                    break;
+                                case "sz":
+                                    props[scopeProperty.Name] = +scopeProperty.RawValue;
+                                    break;
+                                case "s":
+                                    props[scopeProperty.Name] = +scopeProperty.RawValue / 1000000000;
+                                    break;
+                                case "ns":
+                                    props[scopeProperty.Name] = +scopeProperty.RawValue;
+                                    break;
+                                case "ts":
+                                    props[scopeProperty.Name] = new Date(+scopeProperty.RawValue / 1000).toISOString();
+                                    break;
+                                case "cnt":
+                                    props[scopeProperty.Name] = +scopeProperty.RawValue;
+                                    break;
+                                case "cpu":
+                                case "skw":
+                                case "node":
+                                case "ppm":
+                                case "ip":
+                                case "cy":
+                                case "en":
+                                case "txt":
+                                case "id":
+                                case "fname":
+                                default:
+                                    props[scopeProperty.Name] = scopeProperty.RawValue;
+                            }
+                            props["__" + scopeProperty.Name] = scopeProperty.Formatted;
+                        }
+                        data.push({
+                            id: scope.Id,
+                            name: scope.ScopeName,
+                            type: scope.ScopeType,
+                            ...props
+                        });
+                    }
+                }
+                return {
+                    meta,
+                    columns,
+                    data
+                };
             });
         }
 
-        var context = this;
-        this._metricSelectValue = this.metricsSelect.node().value || "TimeElapsed";
-        return this.fetchDetailsNormalizedPromise.then(function (response) {
-            context._graphLookup = {};
-            context._subgraphLookup = {};
-            var data = response.data.filter(function (row) {
-                if (row.type === "graph") context._graphLookup[row.name] = row.id;
-                if (row.type === "subgraph") context._subgraphLookup[row.name] = row.id;
+        return this.fetchDetailsNormalizedPromise.then(response => {
+            this._graphLookup = {};
+            this._subgraphLookup = {};
+            var data = response.data.filter(row => {
+                if (row.type === "graph") this._graphLookup[row.name] = row.id;
+                if (row.type === "subgraph") this._subgraphLookup[row.name] = row.id;
                 if (!row.id) return false;
-                if (context._scopeFilter && row.name !== context._scopeFilter && row.name.indexOf(`${context._scopeFilter}:`) !== 0) return false;
-                if (row[context._metricSelectValue] === undefined) return false;
+                if (this._scopeFilter && row.name !== this._scopeFilter && row.name.indexOf(`${this._scopeFilter}:`) !== 0) return false;
+                if (this._metricSelectValue.every(m => row[m] === undefined)) return false;
                 return true;
             }).sort(function (l, r) {
                 if (l.WhenStarted === undefined && r.WhenStarted !== undefined || l.WhenStarted < r.WhenStarted) return -1;
@@ -170,32 +249,26 @@ export class Timings {
                 if (response.columns[key].Measure && response.columns[key].Measure !== "label") {
                     colArr.push(key)
                 }
-                if (key === context._metricSelectValue) {
-                    measure = response.columns[key].Measure;
-                }
             }
-            context._metricSelectLabel = context._metricSelectValue + (measure ? " (" + measure + ")" : "");
-            var options = context.metricsSelect.selectAll("option").data(colArr, function (d) { return d; });
+            colArr.sort(d3Ascending);
+            this._metricSelectLabel = this._metricSelectValue + (measure ? " (" + measure + ")" : "");
+            var options = this.metricsSelect.selectAll("option").data(colArr, function (d) { return d; });
             options.enter().append("option")
-                .property("value", function (d) { return d; })
                 .merge(options)
-                .text(function (d) { return d; })
+                .property("value", d => d)
+                .property("selected", d => this._metricSelectValue.indexOf(d) >= 0)
+                .text(d => d)
                 ;
             options.exit().remove();
-            context.metricsSelect.node().value = context._metricSelectValue;
 
-            context.chart
-                .columns([context._scopeFilter || "All", context._metricSelectLabel || "Weight"])
-                .data(data.filter(function (row, i) {
-                    return row.name !== context._scopeFilter;
-                }).map(function (row, i) {
-                    return [row.id, row[context._metricSelectValue]];
+            this.chart
+                .columns(["id", ...this._metricSelectValue])
+                .data(data.filter((row, i) => row.name !== this._scopeFilter).map((row, i) => {
+                    return [row.id, ...this._metricSelectValue.map(metric => row[metric])];
                 }))
-                .yAxisTitle(measure)
-                .render()
                 ;
 
-            return data;
+            return [this._metricSelectValue, data];
         });
     }
 

